Ensure WebFullScreenManager is removed as EventListener when page closes

jyavenard opened the pull request:
#### 53ce8c35db4324b63de6d0a600298eb99387395d
<pre>
Ensure WebFullScreenManager is removed as EventListener when page closes
<a href="https://bugs.webkit.org/show_bug.cgi?id=251566">https://bugs.webkit.org/show_bug.cgi?id=251566</a>
rdar://103823114

Reviewed by Youenn Fablet.

When the page close, break the cycles with WebFullScreenManager and ensure
WebFullScreenManager is no longer registered as event listener.

* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
(WebKit::WebFullScreenManager::~WebFullScreenManager):
(WebKit::WebFullScreenManager::invalidate):
(WebKit::WebFullScreenManager::element):
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::close):

Canonical link: <a href="https://commits.webkit.org/259855@main">https://commits.webkit.org/259855@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f78046807e497f68ad4742e6c3eb20040952d661

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106016 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15072 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38849 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115198 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16496 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6252 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98224 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/114933 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111769 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12558 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95528 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40098 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94423 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27171 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81770 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8326 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28523 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8816 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/5082 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14440 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48066 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6812 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10360 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
```diff
@@ -89,9 +89,19 @@ WebFullScreenManager::WebFullScreenManager(WebPage* page)
 WebFullScreenManager::~WebFullScreenManager()
 {
     clearElement();
```
No longer needed since we do it in invalidate().

youennf commented on Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
```diff
@@ -89,9 +89,19 @@ WebFullScreenManager::WebFullScreenManager(WebPage* page)
+#if ENABLE(VIDEO)
+    setMainVideoElement(nullptr);
+#endif
+    m_mainVideoElementTextRecognitionTimer.stop();
```
Should close() call invalidate() as well?
m_mainVideoElementTextRecognitionTimer probably deserves to be stopped when close() is called.

jyavenard replied:
good spotting, can simplify the code

youennf commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -1580,7 +1580,10 @@ void WebPage::close()
 
 #if ENABLE(FULLSCREEN_API)
-    m_fullScreenManager = nullptr;
+    if (m_fullScreenManager) {
```
Will the following work:
```
if (auto manager = std::exchange(m_fullScreenManager, { })
    manager->invalidate();
```

jyavenard committed:
Ensure WebFullScreenManager is removed as EventListener when page closes
https://bugs.webkit.org/show_bug.cgi?id=251566
rdar://103823114

Reviewed by Youenn Fablet.

When the page close, break the cycles with WebFullScreenManager and ensure
WebFullScreenManager is no longer registered as event listener.

* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
(WebKit::WebFullScreenManager::~WebFullScreenManager):
(WebKit::WebFullScreenManager::invalidate):
(WebKit::WebFullScreenManager::element):
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::close):

Canonical link: https://commits.webkit.org/259855@main
